core/src/main/java/org/infinispan/filter/CacheFilters.java

```diff
@@ -80,13 +80,17 @@ private CacheFilters() { }
    public static <K, V, C> Stream<CacheEntry<K, C>> filterAndConvert(Stream<CacheEntry<K, V>> stream,
            KeyValueFilterConverter<? super K, ? super V, C> filterConverter) {
       // Have to use flatMap instead of map/filter as reactive streams spec doesn't allow null values
-      return stream.flatMap(new FilterConverterAsCacheEntryFunction(filterConverter));
+      return stream.flatMap(flatMap(filterConverter));
    }
 
    public static <K, V, C> CacheStream<CacheEntry<K, C>> filterAndConvert(CacheStream<CacheEntry<K, V>> stream,
             KeyValueFilterConverter<? super K, ? super V, C> filterConverter) {
       // Have to use flatMap instead of map/filter as reactive streams spec doesn't allow null values
-      return stream.flatMap(new FilterConverterAsCacheEntryFunction(filterConverter));
+      return stream.flatMap(flatMap(filterConverter));
+   }
+
+   public static <K, V, C> Function<CacheEntry<K, V>, Stream<CacheEntry<K, C>>> flatMap(KeyValueFilterConverter<? super K, ? super V, C> filterConverter) {
```

Comment:
I agree the naming could be better. I am not sure I like the `nullable` part. Maybe `filterConverterToFlatMapFunction` is more descriptive :D

Reply:
My thinking was it's just a function, it's not only usable in `flatMap`, but you're right, we can't focus just on the return type and ignore the function types.

However, that made me look again at the usages of `flatMap()`, and it seems to me that we always create the `KeyValueFilterConverter` instance just before we wrap it with `flatMap()`. Couldn't we simplify things by replacing `IterationFilter` and `CacheEventFilterConverterAsKeyValueFilterConverter` with implementations of `Function<CacheEntry<K, V>, Stream<CacheEntry<K, C>>>`?

The same applies to `function()` (it's used to wrap an unknown `Converter` in `Caches.entrySet()`, but that method is never used and it's not API) and `predicate()` (`Caches.entrySet()` is only called with `AcceptAllKeyValueFilter`).

Thinking further about `flatMap()`, I'm not even sure `flatMap()` is the best `Stream` method to use to handle the `null` values. Because the stream values are actually entries, we can use `NullCacheEntry`, something like this:

```java
   /**
    * Adapt {@link KeyValueFilterConverter} for use with streams by replacing null values with NullCacheEntry instances.
    */
   @Scope(Scopes.NONE)
   static class FilterConverterAsCacheEntryMapper<K, V, C> implements Function<CacheEntry<K, V>, CacheEntry<K, C>> {
      protected final KeyValueFilterConverter<? super K, ? super V, C> converter;

      protected InternalEntryFactory factory;

      public FilterConverterAsCacheEntryMapper(KeyValueFilterConverter<? super K, ? super V, C> converter) {
         Objects.requireNonNull(converter);
         this.converter = converter;
      }

      @Inject
      public void inject(InternalEntryFactory factory, ComponentRegistry registry) {
         this.factory = factory;
         registry.wireDependencies(converter);
      }

      @Override
      public CacheEntry<K, C> apply(CacheEntry<K, V> kvCacheEntry) {
         K key = kvCacheEntry.getKey();
         V value = kvCacheEntry.getValue();
         Metadata metadata = kvCacheEntry.getMetadata();
         C converted = converter.filterAndConvert(key, value, metadata);
         if (converted == null) {
            return NullCacheEntry.getInstance();
         }
         return factory.create(key, converted, metadata);
      }
   }

   public static <K, V, C> CacheStream<CacheEntry<K, C>> filterAndConvert(CacheStream<CacheEntry<K, V>> stream,
                                                                          KeyValueFilterConverter<? super K, ?
                                                                                super V, C> filterConverter) {
      return stream.map(new FilterConverterAsCacheEntryMapper<>(filterConverter))
                   .filter(c -> c != NullCacheEntry.getInstance());
   }
```

Even when `MortalCacheEntry` is not available, it may be cheaper to use a `new Object()` as a null marker to get less allocations (at the expense of more lambdas):

```java
   public static final Object NULL_MARKER = new Object();

   @Scope(Scopes.NONE)
   static class FilterConverterAsValueMapper<K, V, C> implements Function<CacheEntry<K, V>, Object> {
      protected final KeyValueFilterConverter<? super K, ? super V, C> converter;

      public FilterConverterAsValueMapper(KeyValueFilterConverter<? super K, ? super V, C> converter) {
         Objects.requireNonNull(converter);
         this.converter = converter;
      }

      @Inject
      public void inject(ComponentRegistry registry) {
         registry.wireDependencies(converter);
      }

      @Override
      public Object apply(CacheEntry<K, V> entry) {
         C converted = converter.filterAndConvert(entry.getKey(), entry.getValue(), entry.getMetadata());
         if (converted == null) {
            return NULL_MARKER;
         }
         return converted;
      }
   }
   
   public static <K, V, C> CacheStream<C> filterAndConvertToValue(CacheStream<CacheEntry<K, V>> stream,
         KeyValueFilterConverter<? super K, ? super V, C> filterConverter) {
      // Have to use flatMap instead of map/filter as reactive streams spec doesn't allow null values
      return stream.map(new FilterConverterAsValueMapper<>(filterConverter))
            .filter(Objects::nonNull)
            .map(c -> (C) c);
   }
```

I'm not sure how it would work out for RxJava and `Maybe`, but I did notice some usages of `RxJavaInterop.completionStageToMaybe()` that can be simplified, either by returning `null` and removing the additional `filter()` step, or by replacing it with `completionStageToSingle()` and removing the additional `toSingle()` step.

Reply:
> Thinking further about `flatMap()`, I'm not even sure `flatMap()` is the best `Stream` method to use to handle the `null` values. Because the stream values are actually entries, we can use `NullCacheEntry`, something like this:

This was my idea originally, but something prevented it a while back. My guess is this is probably fine now though. The only problem is that these methods are public, so I will have to make some new methods instead. I will mess with it a bit though.

And I agree that `IterationFilter` can be redone, but I figure that is a bit out of scope of this PR at this time.